Make UriTranslation check prefix contents before matching

aeons opened the pull request:
This PR changes the behaviour of `UriTranslation` to take the contents of the prefix into account as well.

It inlines the old implementation in `Router`, to not do the check twice.

Fixes #1962

aeons commented:
I'm not completely clear on the semantics of UriTranslation.

Should it just ignore whether there's a slash at the front of the prefix and request?

ChristopherDavenport commented on server/src/main/scala/org/http4s/server/middleware/URITranslation.scala:
```diff
@@ -2,21 +2,33 @@ package org.http4s
-      case x if x.startsWith("/") => x.length
-      case x => x.length + 1
-    }
+  def translateRoot[F[_]: Applicative](prefix: String)(service: HttpService[F]): HttpService[F] =
```
This changes the function signature, so isn't binary compatible

aeons replied:
Hmm, yes, that's true.
I'm not sure if this can be solved without being to signal a not-matched via `OptionT.none`.

aeons replied:
@rossabaker could you take a look at this? Should I retarget master and just let this be a known bug in 0.18?

rossabaker replied:
Our type is called `Uri`, but this object is called `URI`.  That's inconsistent.  We could deprecate thid old buggy one, and use this new implementation as `UriTranslation` in release-0.18.x.  Or `TranslateUri`.

aeons replied:
:+1:

aeons commented:
I rebased and changed the following:

- Readd deprecated `URITranslation`
- Add the new implementation as `TranslateUri.apply` (I got a warning about only differing in case if it was called `UriTranslation`)

ChristopherDavenport commented on server/src/main/scala/org/http4s/server/middleware/URITranslation.scala:
```diff
@@ -4,6 +4,7 @@ package middleware
 
 import cats.Functor
 
+@deprecated("Use org.http4s.server.middleware.UriTranslation instead", since = "0.18.16")
```
Do you mean `TranslateUri`?

rossabaker replied:
Good catch.  If only we could compile our deprecation messages.

aeons replied:
Yeah, I tried just changing the case, but got a compiler warning.
